Initial implementation of type stubs (mypy/PEP484)

ksunden opened the pull request:
## PR Summary

Closes #20504 

This is a massive PR and I realize that... I don't fully expect anyone to read every single line of the newly added stub files. That said, these files have zero effect at run time for the library, so even if there are things that are wrong about them, they won't cause running code to fail. I'd like to get this in early in the 3.8 cycle to allow downstream libraries to test out the type hints with sufficient time to fix them before final release.

When reviewing, read enough of the stub files to get a sense of how they work, and to understand common patterns.
I attempted to be as complete and correct as I could be, but that is mostly reliant on existing docstrings being complete and correct (and some amount of fallback to variable names).

Highlighted changes to library code:
- `__all__` added to `matplotlib.__init__` 
    - `mypy` was not happy about some things being implicitly exported (things that were imported but other places use in the mpl top level namespace)
    - This makes explicit that all mpl defined things are publically accessible from the mpl namespace
    - Happy to add/subtract from the list
- Pyplot has type hints inline
    - boilerplate.py updated to gather annotations
    - because of the syntax rules of the return annotation (cannot line break on ` -> `) the textwrap based line wrapping has been removed and replaced with a clean-up step after writing of running `black`
    - This makes some of the other behaviors (such as putting null bytes to avoid line wrapping the data kwarg) obsolete, but at the cost of a new dev (and testing) dependency on black
    - The handwritten portions of pyplot are wrapped with `# fmt: off/on` comments, so only the autogenerated portion of the file is actually run through `black`
- Some type aliases for commonly used types:
    - Names are negotiable, mostly just tacked on `Type` to the idea I was representing to avoid occluding a name we might want to use
    - These are included in the .py files such that downstream users can use them as type hints
      - In the earlier days of this work, they were in a `_typing.pyi` stub file that only worked while type checking, but redistributed to more natural homes and avoided that complexity.
    - `colors.Color` (str or 3/4-tuple of floats)
      - may wish to rename, especially if we want to be able to create a real `Color` class to handle things like comparisons, which has been proposed recently.
    - `markers.MarkerType`(str, Path, MarkerStyle), `markers.FillStyleType` (Literal of strings)
    - `lines.LineStyleType` (str or tuple for dashes), `lines.DrawStyleType` (Literal of strings), `lines.MarkEveryType` (many types possible)
    
    

For most of the library, stub files are used rather than inline type hints.
This has the advantage of not touching literally every public signature line in one PR, but does carry a tradeoff of direct utility of `mypy` and some other tools in the typing space.
`mypy` will trust the pyi files and disregard the actual implementation, meaning that if a parameter is added/removed, but the pyi file is not updated, mypy will use the outdated signature to type check.

The two exceptions where inline type hints are used are tests and pyplot.
Tests use inline typehints, but do so relatively minimally, only when mypy complains explicitly. It doesn't make a lot of sense to write stub files for tests, which are not called by downstream libraries (who the stub files are actually mostly for) but running type checking on tests makes a fair amount of sense as it is more representative of what downstream users will see.

Pyplot uses inline type hints because it is autogenerated, and it was easier to inject the type hints into the existing autogenerated code than it would be to do parallel lookup and write two files. The hand written portions of pyplot also do have type hints.
Since pyplot has inline type hints, mypy will actually check the implementation of the functions for type consistency, which can find certain errors.

There are some tools to help discover those problems, but I have found that they are too noisy and not configurable enough to be useful to us (yet).
See [this comment](https://github.com/ksunden/matplotlib/pull/1#issuecomment-1374282552) for a round up of some of the other type checking tools that are available and why I don't fully support them yet.
I may end up writing a script to find the most commonly anticipated discrepancies, but for getting started, the heuristic is "if a signature is touched, make sure to update the pyi file" (which are already a higher bar of review/etc as anything that changes a signature is an API change which should carry a release note.)
Most specifically, added/removed functions/methods which are not reflected in the pyi and signature changes.

There are some areas left for followup PRs as they are a) relatively complicated ideas that should get particular attention on review that would be lost in this massive PR or b) ideas that can be added incrementally 

- data arg preprocessor
- typing shapes and dtypes of arrays (most everything is [ArrayLike]( https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike) or np.ndarray, but these types do support being more specific. Using the full power that seems to be available is not as well documented as I would have liked, so may wish to reach out to somebody who knows more about this (and encourage/possibly write the docs that I think should exist for typing with numpy)
- Some methods (e.g. `tri*`, `streamplot`, etc) are defined outside of the Axes class, and so the pyplot generation code does not currently get their type hints. The code could be updated to be smarter about looking at attributes and getting their type hints, but that would make it more complicated, so cover the 90% case first.

Additionally, we may wish to be more relaxed or more specific in certain areas.

[Dask](https://github.com/dask/community/issues/255) has an interesting set of guidelines that perhaps we should adopt some or all of.
In particular, for this pass I was being as complete as I could, but they recommend things like "if the return type is a union of things, don't actually type it, because that can be more burdensome to downstream users than just saying `Any`"
Also some places where we specify callback functions may be overly specified or hard to follow with the full specification, so relaxing the type hint may aid readability.

I certainly don't follow every one of those guidelines yet, but I think they are mostly well reasoned and perhaps we should adopt them/go through and clean up (things like the return annotation on `__init__` or importing `collections.abc` over `typing` may be better as followup PRs.

I make fairly heavy use of [`typing.Literal`](https://docs.python.org/3/library/typing.html#typing.Literal). There may be some places where that is more unwieldy than we'd like and it should be `str`, but for the most part it has a reasonable correspondence to where we use `_api.check_in_list` (and its similar siblings). There may also be places where I went more generic but we can be more specific by using Literal.

Some types, particularly things like `tuple[float, float, float, float]` are a little unwieldy and may be better understandable replaced by a type alias (e.g. `RGBA` or `RectangleSpec` (names negotiable), because despite being the same type, their usage is logically different).
Also, tuple types are sometimes used when what we really mean is "unpackable thing with N elements" but the type system doesn't handle that _super_ well at this time (though perhaps ArrayLike can get us there...). I also think it is fair in those cases that we say "you _should_ give us a tuple, and for the best guarantee of forward compatibility, maybe cast it to one just to be sure".

The sphinx build is _mostly_ unaffected, as sphinx does not read pyi stub files (yet)... though pyplot does get all of the info in the signature lines, and required some things to be added to the list of ignored warnings (for now, at least... perhaps we could improve their ability to link at some point)

Todos before merge:

- [x] Move items from `_typing.pyi` to their more natural homes, and document them. This was a bit of a catchall, that ended up not having all that many things in it, but I wanted to be able to use the types and move them later. It is part of why sphinx is unhappy, since these are not documented.
- [x] Fix sphinx build
- [ ] Add documentation, especially of expectations moving forward (e.g. keeping these up to date)
- [ ] Write release note[s]


## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

ksunden commented:
huh, odd, I don't get those sphinx warnings locally... not a huge deal, just odd

QuLogic commented on lib/matplotlib/pyplot.py:
```diff
@@ -1479,10 +1563,19 @@ def subplots(nrows=1, ncols=1, *, sharex=False, sharey=False, squeeze=True,
-                   subplot_kw=None, gridspec_kw=None,
-                   per_subplot_kw=None, **fig_kw):
+def subplot_mosaic(
+    mosaic: str | list[list[Any]],
```
According to https://stackoverflow.com/a/53845083, you should be able to annotate the recursive list now (the linked mypy issue there is fixed.)

ksunden replied:
I'll try it out, though I suspect it may be harder to parse as a human than just seeing `Any` plus the docstring (and will require some additional named types, since you can't do a forward reference inline)

ksunden replied:
It does in fact work, and isn't as hard to read as I feared:

```python
    HashableList = list[Hashable | "HashableList"]
    def subplot_mosaic(
        self,
        mosaic: str | list[HashableList],
```

(tested in the `figure.pyi` version of this function)

It does require the stringified forward reference, which is otherwise mostly not needed in `pyi` files or `py` files with the `__future__` import, so there's a mild confusion potential from that, but not actually that bad.

Holding off until committing this as `HashableList` is something that might go in the `mpl.typing` module as it is purely for typing purposes.

QuLogic commented on lib/matplotlib/pyplot.py:
```diff
@@ -2164,16 +2296,18 @@ def set_cmap(cmap):
 
 @_copy_docstring_and_deprecators(matplotlib.image.imread)
-def imread(fname, format=None):
+def imread(fname: str | io.FileIO, format: str | None = None) -> np.ndarray:
```
Should have `os.PathLike` as well?

ksunden replied:
Possibly, I think it may work.

That said, the docstring is listed as `str or file-like` and there are `isinstance(fname, str)` checks in the implementation, so I kept it as the narrower type.

My inclination is to leave it for now, but maybe do a narrower change that enables any `os.PathLike` and ensures that the implementation matches, the docstring is updated, and the type hint is updated both in `image.pyi` and in `pyplot.py` (this is not an autogenerated method)

tacaswell replied:
we should support PathLike everythere (even if the docs did not catch up).

ksunden replied:
PIL doesn't actually cover the full `os.PathLike` (though the common cases of `bytes`, `str`, and `pathlib.Path` are covered)

They use the following check:

https://github.com/python-pillow/Pillow/blob/583dd3b4245649ddd367c2dcbbf2bf9fb7835017/src/PIL/_util.py#L5-L6

QuLogic reviewed (commented):
I've only looked at the `.py` files and lightly skimmed the `.pyi` to get a general feel for things.

QuLogic commented:
> I make fairly heavy use of [typing.Literal](https://docs.python.org/3/library/typing.html#typing.Literal). There may be some places where that is more unwieldy than we'd like and it should be str, but for the most part it has a reasonable correspondence to where we use _api.check_in_list (and its similar siblings). There may also be places where I went more generic but we can be more specific by using Literal.

I wonder if `_api.check_in_list` should be changed/reproduced as a decorator that can check the `Literal` signature to avoid duplication. Or perhaps that it too much work.

ksunden commented on lib/matplotlib/pyplot.py:
```diff
@@ -1125,7 +1201,7 @@ def cla():
 
 @_docstring.dedent_interpd
-def subplot(*args, **kwargs):
+def subplot(*args, **kwargs) -> matplotlib.axes.Axes:
```
This was admittedly an attempt to get sphinx to be quieter before I just ignored the warnings (`Axes` has multiple targets, and sphinx doesn't resolve which one is actually used, it just looks at the string)

It can (And probably should) be cleaned up and just ignored in sphinx

ksunden replied:
Well, looks like my ignores didn't work anyway, those only work for missing references, not ambiguous ones.

So I added it back as `axes.Axes` to quiet sphinx warnings. Not as pretty, but better unwilling to turn of warnings -> errors in sphinx, so perhaps the best I can do at this time.

For reasons unclear to me, the some references do _not_ generate this warning, such as `sca`...

ksunden replied:
well `axes.Axes` is no good because `axes` is defined in pyplot, so using the full `matplotlib.axes.Axes` instead, which is even more ugly, but best I've got

ksunden commented on lib/matplotlib/__init__.py:
```diff
@@ -103,6 +103,36 @@
 
 """
 
+__all__ = [
```
[not according to PEP8](https://peps.python.org/pep-0008/#module-level-dunder-names), `__all__` is explicitly 'supposed' to go after only `__future__` imports, but before all other imports (technically would also apply to `__bibtex__` and `__version[_info]__`, but the former is a relatively rare dunder and the latter has computation involved, so not much to do there)

ksunden commented on tools/boilerplate.py:
```diff
@@ -379,6 +382,78 @@ def build_pyplot(pyplot_path):
         pyplot.writelines(pyplot_orig)
         pyplot.writelines(boilerplate_gen())
 
+    # Run black to autoformat pyplot
+    subprocess.run(
+        [sys.executable, "-m", "black", "--line-length=79", pyplot_path]
+    )
```
Certainly it is possible, but the tradeoff would be that the generation code would be less readable (and possibly the output code as well, though in a more subjective way)

If I were to write it, I would probably at least start with black-style (even if generated ourselves) for the signatures, rather than using `textwrap` (but perhaps keeping `textwrap` for the method bodies)... with type hints each variable (and the return annotation ) being on a separate line is much more readable, as it is often less obvious where a new variable starts, especially with types that include commas. (I would probably still try to compress to a single line if it'll fit, but that is relatively rare with the added length of the annotations).
While relatively rare, "one line per parameter" is not always itself sufficient to guarantee short enough lines, so that requires additional handling.
I'm not over the moon about how `black` handles these cases. I think it is a weak point of `black`'s style. But it's probably still better at handling these edge cases than most formatting code that is itself readable/short enough to not be totally lost in formatting weeds.

The fact of the matter is that I was apprehensive about writing it even with the list of edge cases I saw quickly, and I'm sure I would discover at least a few more while implementing. `black` was quite honestly the quickest answer to the problems I had at hand, so I went with it.

I don't think the bump to 88 will significantly impact whether `textwrap` will do the trick... there are far too many of these calls that span multiple lines and so while it will shift which ones get cut off on the ` -> `, I doubt it will get rid of them (and certainly even if we get lucky, it will be fragile to future change)
It _is_ more likely to ensure that more of the type hints fit on a single line, but still won't get some of the larger Unions/parameters with `Callable` type hints.
I can also think of other ways `textwrap` is fragile that are more likely to be problematic with type hints. For example, if a string with a space appears in the signature (not something we commonly do, but some of the positioning code for things like legend take strings like `"upper right"`, so if we decided to encode that into a Literal type or use that as the default, there would be nothing preventing the `textwrap` based code formatting from breaking on that space, which is not syntactically valid.
We could probably get away without addressing these edge cases for quite a long time, but eventually are likely to need something more syntax aware in more and more ways.


In any case, the move to `black` for pyplot was primarily one of not wanting to get lost in minutia of how autogenerated code is formatted while retaining correct syntax. 
I think the generation code is easier to understand, as I was able to get rid of some of the other workarounds (like using a `\0` to avoid breaking on the data arg dict unpacking, which I could also just do something similar for ` -> ` if my only goal was "<X characters and syntactically valid")

It now has in the generation code with only the "syntactically valid" code and then a clean up at the end to do both "<X characters" and "mostly readable", without breaking "syntactically valid". Those are precisely the _goals_ of `black`(we can argue about the degree to which it achieves it, but it is the goal).
While readability is _less_ useful for autogenerated code than handwritten code, I'd argue that it still matters, especially for signatures.


Wow, that got away from me a bit more than I expected when I started writing... tl;dr yes black is perhaps overkill, but type hints complicate signatures in ways that invite formatting edge cases, so I expect any solution that doesn't involve an extra dependency to be more fragile and harder to follow and update the generation code.

ksunden commented on tools/boilerplate.py:
```diff
@@ -117,6 +119,17 @@ def __repr__(self):
         return self._repr
 
 
+class direct_repr:
```
I was style matching to `value_formatter` which does similar repr tricks. not opposed to camel-casing, just was doing similar things to existing code.

tacaswell replied:
Given that this is internal code, I think we should merge as-is and then clean up both later.

ksunden commented on tools/boilerplate.py:
```diff
@@ -379,6 +382,78 @@ def build_pyplot(pyplot_path):
+    path = Path(inspect.getfile(cls))
+    stubpath = path.with_suffix(".pyi")
+    path = stubpath if stubpath.exists() else path
+    tree = ast.parse(path.read_text())
```
I wasn't overly concerned with optimizations, but indeed adding `functools.lru_cache` to `get_ast_mro_trees` (which is only ever called on Axes and Figure, and is the method called in the body of the generation) does speed it up by 30% or so (3 seconds to 2 seconds on my machine, including the subprocess call to black)

Adding an additional cache to this function would basically only prevent `Artist` from being parsed twice.

oscargus commented:
Now it is probably not worth it, but I still would like to point out https://github.com/microsoft/python-type-stubs/tree/main/matplotlib that you may or may not be aware of...

ksunden commented on lib/matplotlib/colors.py:
```diff
@@ -56,6 +58,12 @@
 from matplotlib import _api, _cm, cbook, scale
 from ._color_data import BASE_COLORS, TABLEAU_COLORS, CSS4_COLORS, XKCD_COLORS
 
+Color = Union[
```
I can certainly see an argument in either direction here... I initially started with a `_typing.pyi` file that I filled with a handful of type definitions that were used in multiple files.

When I was done I somewhat felt that the definitions that were there (~6 of them, highlighted in the third bullet point above) were narrowly scoped enough that placing them together instead of with their use cases was a little awkward, but I could certainly see doing a `typing.py` instead of where I ended up.

I do think that it should be public (and available at runtime so that downstream can type check without necessarily doing `if TYPE_CHECKING:` blocks) rather than what I had initially done. But it can be more explicit about being typing.

the `_typing.pyi` file has been wiped from the git history since it was both added and deleted and was triggering the Cleanliness check to fail.

Even if we don't end up moving to a `typing` module, I was considering whether `Color` in particular should be `ColorType` or similar. Especially given conversation in #24859 about potentially creating an actual `Color` class to allow some level of comparisons between string forms and RGB(A) forms of colors.

tacaswell replied:
I am :+1: on calling this `ColorType` (`ColourType`)?

ksunden commented on lib/matplotlib/pyplot.py:
```diff
@@ -1108,15 +1184,15 @@ def delaxes(ax=None):
     """
     figure(ax.figure)
-    ax.figure.sca(ax)
+    ax.figure.sca(ax)  # type: ignore
```
I want to say its the minor difference between `_AxesBase` and `Axes`, but I'll make sure and either remove the type ignore or add a more helpful comment

ksunden replied:
It was actually that `ax.figure` is typed as `FigureCanvasBase | None`, and None (of course) does not have an `sca` method.

But if you are calling `plt.sca` with an axes not in a figure already, that is its own problem.

The alternative would be to raise `if ax.figure is None:` which would I think quiet mypy because of type narrowing.

ksunden replied:
added the aforementioned comment about what was being ignored in d8329509

ksunden commented on lib/matplotlib/__init__.pyi:
```diff
@@ -0,0 +1,115 @@
+__all__ = [
```
Technically true that mypy doesn't need it in the `py` file at all (it only looks at the `pyi`). But if we can be a tad more specific about what should be relied on as importable from that location, I'd lean towards it, personally.

The other (in my opinion ugly) answer is to put aliases for all of the things we want exported or to enable the [mypy option](https://mypy.readthedocs.io/en/stable/config_file.html#confval-implicit_reexport) for implicit exports.

The former means a lot of `import bar as bar` (which stubgen does by default, but I got rid of most of the time)

ksunden commented on lib/matplotlib/pyplot.py:
```diff
@@ -1805,7 +1903,13 @@ def ylim(*args, **kwargs):
 
-def xticks(ticks=None, labels=None, *, minor=False, **kwargs):
+def xticks(
+    ticks: Sequence[float] | None = None,
```
Ultimately, this comes down to `axis.Axis.set_ticks` being documented as `list of floats`... I apparently translated that to `Iterable[float]` there(/in Axes), and for some reason slightly narrowed back to `Sequence[float]` for pyplot. The MS type stubs just typed it as `ArrayLike`, which is broader.

Units are a particular case for typing that I haven't gone through and found all of the places where they would apply/expand types. This was partly because they were not being noisy in mypy and partly because I see that as an easier to review change as a perturbation rather than included in full here. (units.py and associated files like dates and category are not actually type hinted yet at all, they don't have any warnings raised and so are separable in that sense, though of course do interweave in many places in Axes/Axis)

So in short, even if it is technically not 100% correct, I think this is a change better made as a follow up. It is consistent with docstrings, so if it should change, the change is likely more than _only_ type hints. (I've been trying to keep this PR hemmed in by that where possible, though a handful of places where mypy warns were addressed)

ksunden commented on lib/matplotlib/pyplot.py:
```diff
@@ -797,7 +867,8 @@ def figure(num=None,  # autoincrement if None, else integer from 1-N
         num = int(num)  # crude validation of num argument
 
-    manager = _pylab_helpers.Gcf.get_fig_manager(num)
+    # Type of "num" has narrowed to int, but mypy can't quite see it
+    manager = _pylab_helpers.Gcf.get_fig_manager(num)  # type: ignore[arg-type]
```
This falls into the category of things where the implementation certainly allows it to some degree, but I'm curious about the reasons why it is actually allowed/whether it is actually encouraged... it is referred to as "number" everywhere except for one parenthetical in a docstring...

And if we look deeper than "these are the keys of an internal dictionary", it gets murkier:

The fact that just above this we use `int(num)` as a "crude validation" if num is not a str/None makes me think that it actually _doesn't_ fully work with non integer values, at least not as expected, for figures generated through `plt.figure` and if that's the case, I think the course of action is to remove the parenthetical from the docstring... (the Gcf class in and of itself, certainly does work with any hashable, but it is a class defined in a private module and only really used for the purposes of `pyplot`, so if it doesn't work in that context, I'm not sure why it should be documented to work at all)

For that matter, the implementation of `plt.figure` would lead me to believe it only works properly if the labels are specifically `str` not `Any`...

ksunden commented on lib/matplotlib/_pylab_helpers.pyi:
```diff
@@ -0,0 +1,27 @@
+from matplotlib.figure import Figure
+
+class Gcf:
+    figs: OrderedDict[int, FigureManagerBase]
```
See comment in pyplot about the Gcf class

ksunden commented:
@oscargus I was aware, they were a valuable resource in this exercise, though are not fully complete nor correct (I believe many of the hints in that repo are extracted from inspecting running code, but there are many cases where the type is wider than it saw, e.g. zorder can be any float, but is most often set to int)

My first attempt was to simply start with those, but that had enough warnings so as to be overwhelming, so I started by doing minimal changes to make mypy happy enough to start and adding single files at a time (that is actually why the branch is called `mypy-gradual` because my first one was just `mypy`, but I wanted to take it slower/more methodically).

ksunden commented on .github/workflows/reviewdog.yml:
```diff
@@ -38,6 +38,40 @@ jobs:
+      - name: Set up Python 3
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.8
```
```suggestion
          python-version: 3.9
```


This fails on merged version of code because `functools.cache` was new in 3.9

ksunden commented:
> I wonder if `_api.check_in_list` should be changed/reproduced as a decorator that can check the Literal signature to avoid duplication. Or perhaps that it too much work.

I think that may be interesting longer term... though probably not as a straight replacement.

- Cases where `check_in_list` include dynamic usage which does not work for type hints (e.g. register)
- Probably requires inline type hints (as the pyi type hints are not included at runtime)

ksunden commented:
I have added at least some documentation, though perhaps more is warranted (though that may be follow up as we start to get used to using this and know more about what needs to be documented).

From my perspective the biggest questions prior to merge are:

- Do we want a `matplotlib.typing` module? (where type aliases/unions/perhaps some literals get shoved)
- What are the feelings about "figure number" which are not numbers and can/should type hints narrow that expectation?

I think there are the following tasks, which are relatively uncontroversial but relate to the `mpl.typing` module if it exists:

- subplot mosaic `HashableList` type hint
- rename `Color` to `ColorType` (possibly with alias `ColourType` if we so desire)

The latest commit fixes the recent mypy identified errors by adding additional error handling.

ksunden commented:
Decisions re open questions above:

 - Yes do a `typing` module as a public module, but be very clear about provisional status and reserve right to change at any time without warning or deprecation (in API docs and release notes).
 - Retain stricter type hints for Gcf to start, if they need to be expanded because a usecase is found and advocated for, we can do so at that time.
 - Ensure mypy is happy on 3.9 (revert/add `__future__` import changes that affect 3.9)
   - remove explicit `TypeAlias` (3.10+ feature, leave comment in `typing.py` to add when 3.9 is dropped)
   - Ensure `__future__` import works for `|` and `type[]` notations

oscargus commented:
As I've just looked into typing for another project: how does this work out with pyright? My experience from the other project is that pyright is somewhat easier to make happy... Not that it should be an argument to select one, but it would be nice if it worked well with both (as pyright is more often used for type hinting in IDEs if I am not wrong).

ksunden commented:
For reference, pyright with a similar configuration to to mypy (i.e. excluding the same subfolders, but using defaults otherwise) finds 10752 errors, when mypy finds 0.

If I additionally exclude the `tests` subfolder, that does go down to 2998 errors.

Many of these are probably do to overly narrow inferred types in code that we do not provide explicit type hints for.
I think if we more fully embrace type hints going forward for internal code (the stub files are still primarily for external/downstream code moreso than for ourselves) then I think addressing these (perhaps more slowly or something like GSOC in a future year) would likely be beneficial for maintaining consistent typed code over time.

There are likely configuration options that could narrow down to a more manageable subset and then incrementally add from there.

I guess my opinion is pyright may be easier to keep happy in the extreme of "everything is typed", but I don't think it is easier in the "I'm adding type hints to a larger extant codebase"

ksunden commented:
the script I added to tools essentially is doing a subset of what mypy's `stubtest` does, but I could not configure stubtest to be less noisy.

ksunden commented:
Rebased after merging #24984, discovered my tool was accidentally ignoring `__init__` (because of the leading underscore)

Updated and corrected all of the `__init__` discrepancies.

ksunden commented:
Okay, sitting down with mypy's `stubtest`, I think working towards using that as the CI check is preferable to maintaining/adding in many more edge cases to my own tool, but not quite sure on what timescale.

I'm still not super clear on how the `--allow-list` argument works for that tool, I added it and still had many warnings from things I had removed from my allow list (and the docs pretty much just say "use the allow list to ignore things")

I do not think there is a complimentary "ignore-list" idea, which may be preferable, especially as it comes to some of the reasons we may wish to ignore, outlined below:

It has some poor interactions that need to be worked around:

It opens one interactive figure during analysis, which is odd, but adding an envvar fixes that much pretty easily.

~4k of ~8.7k errors are explicitly in non-typed code (e.g. backends, tests, pylab being the largest offenders)

~3.8k of the ~4.5k remaining errors are all to do with `Artist.set` methods, largely because subclasses use `**kwargs` instead of enumerating the precise (large) set of arguments in every implementation.

Many of the remaining ~700 errors are poor interactions with things like our deprecation machinery (especially renaming or morphing into kwonly) as `stubgen` (unlike my tool) uses imported versions (which I think is a good choice, for many reasons, but has implications for platform specific code, etc). Also included in this category is the handling of `data` argument for artist functions.

So in summary, the rather overwhelming initial number of almost 9 thousand errors is cut down to a much more manageable number pretty quickly, and if I could work out how to actually ignore them (rather than using grep post hoc as I have been for this analysis). I think my tool may be a stopgap CI check still while the wrinkles get ironed out.

ksunden commented:
Categories of errors from examples:

- stem from lack of type hints in the example file
  - e.g. overly strict inferred type or reuse of variable for different "cells"
  - not super interested in adding type hints to these files (at least not piecemeal... if we want to commit to type hinted code being recommended, I could see it, but while its still provisional/power user feature, not sure)
- Advanced reliance on specific return type not guaranteed by the type hint
  - e.g. blitting relying on an Agg figure canvas, while the type hint is just FigureCanvasBase
  - e.g. knowing that a specific colormap is a `LinearSegmentedColormap` rather than `Colormap`
  - Sometimes we the programmer know a more narrow subtype will be returned, but it is not available to the type checker (e.g. relies on runtime environment or _values_ (not types) of arguments)
  - More common cases may be better served by relaxing type hinting to `Any`, but perhaps not all
- using `mpl.cm.<colormapname>`
  - These are set by updating `globals` and are not included in the static type checker
  - It is commented as being an implementation detail and that public code should use `mpl.cm["<colormapname>"]`, so I think it is fair to expect type hinted code to use that preferred mechanism.
- passing list literals to functions type hinted as tuples
  - arguably should be allowed, but lose info when relaxing that type hint, tuples have fixed size and potentially different types for each entry. lists are homgeneous and do not have a size inspectable by the type checker.
  - lean towards enforce tuples for type checked code, even if the implementation can take more
- `set` methods (broadly) being a strange edge case for type hinting
  - some of it is the same "I know what subtype I have" as above, but may wish to give up on properly type hinting these methods.
  - This overlaps with `stubtest`s complaints of the same 

This list was curated by looking through the `tutorials`, but I expect the same will be said for `examples`, just there is more of them

I did find a number of inaccuracies by running those, so I think it is a good exercise, but not ready to add to CI, as doing so would likely require adding extensive type hints and/or type:ignore comments to the examples.

ksunden commented:
Okay, I have updated all of the library type hints I intend to without further prompting.

There are some things that may need to be relaxed/etc, but I feel that can be done as a follow on. but most of the remaining errors in the galleries fall into the buckets described above.

One additional category is using deprecated/private attributes, but that is a more temporary problem as those (well, especially the deprecations) will be removed for other reasons (and we _could_ type hint them if we wanted to, just decided it wasn't worth it up front).

ksunden commented:
Test errors caused by `@overload` of `[Axes|Figure].legend` and interactions with `boilerplate.py`.

Namely in relation to adding `**kwargs`, which was found to be missing when analyzing examples.
Only adding `**kwargs` to each signature obscures one of the signatures (and the type checking of keyword-passed args in general).

Fix was to move the `def legend(self, **kwargs)` signature to the bottom but add `def legend(self)` as the first overload entry to avoid breaking `boilerplate.py`, which does not like having type hinted variables that do not appear in the implementation signature, and only looks at the first overload (it could(/should?) be made smarter in that regard, but the workaround was not hard).

ksunden commented:
I decided to do one more rebase post merging of #25456 and in the mean time have actually figured out configuring of mypy's stubtest, so went ahead and added a simple ci run.

The short version of my confusion with configuration was that I had interpreted "allowlist" to mean "allow the checker to check these items", when it actually means "allow these errors to exist without printing/non-zero exit". so it was precisely what we needed, just odd terminology.

Note that unlike some other linters (including mypy itself), we do not get line number info and thus will not get inline reviewdog flags (which would generally be _better_, though even if that info was there, would require a parser for reviewdog).

ksunden commented:
Hmm... the new CI check appears to be ignoring the `exclude` portion of the pyproject.toml mypy config (and thus failing on running mypy, rather than on the following stubtest)

Not sure why it is doing that, cannot reproduce locally... going to set it aside for a bit and come back to it with fresh eyes later.

bschnurr commented:
Could you add a return type annotation to `def figure(..)`?
https://github.com/matplotlib/matplotlib/blob/161c96c1f08282d3d412f313240b11f781daa9de/lib/matplotlib/pyplot.py#L680

ksunden commented:
@bschnurr Unfortunately it is not as simple as just adding the return type hint of `Figure` here... 

Since pyplot is typed inline, the implementation is type checked (unlike the files where we use [yi stubs), and some of returns have looser type hints (namely `FigureBase | None` instead of `Figure`), and that None is the real sticky bit, as it introduces places where calling functions would need to account for the possibility of None, even if they have other reasons for know it is not None...

I'm trying to as much as I can to not touch the implementation in this already massive and hard to review PR (there are a few bits where I added some `isinstance` checks or similar, but trying to limit that as much as possible). Doing so limits the risk of unintended breakage from such a large PR (almost all of this PR is in (new) pyi files, which do not affect run-time behavior at all).

So, this is an area where the trivial thing doesn't quite work and may need to be addressed in the implementation, and thus it should be a candidate for a followup pr that is more narrowly focused (and thus easier to review)

To be clear, I think adding it is possible and not _too_ much additional code, just think that it is tricky enough that I want it done as a follow up instead of rolled in here.

The following diff _almost_ passes mypy (there are a couple errors from the slight difference between `FigureBase` and `Figure` that neither type hint quite nails, and would need a little more thought into whether we can actually guarantee proper `Figure` or not...):

```diff
--- b/lib/matplotlib/pyplot.py
+++ b/lib/matplotlib/pyplot.py
@@ -756,7 +756,7 @@ def figure(
     FigureClass: Type[Figure] = Figure,
     clear: bool = False,
     **kwargs
-):
+) -> FigureBase:
     """
     Create a new figure, or activate an existing figure.
 
@@ -859,7 +859,7 @@ default: None
     in the matplotlibrc file.
     """
     if isinstance(num, FigureBase):
-        if num.canvas.manager is None:
+        if num.canvas.manager is None or num.figure is None:
             raise ValueError("The passed figure is not managed by pyplot")
         _pylab_helpers.Gcf.set_active(num.canvas.manager)
         return num.figure
@@ -2347,7 +2347,7 @@ def matshow(A: ArrayLike, fignum: None | int = None, **kwargs) -> AxesImage:
         # Extract actual aspect ratio of array and make appropriately sized
         # figure.
         fig = figure(fignum, figsize=figaspect(A))
-        ax = fig.add_axes([0.15, 0.09, 0.775, 0.775])
+        ax = fig.add_axes((0.15, 0.09, 0.775, 0.775))
     im = ax.matshow(A, **kwargs)
     sci(im)
     return im
```

The three changes here are:

- Add the type hint itself (I went with FigureBase in this instance, but see discussion below)
- Account for the `Figurebase.figure` being None
   - I think this is only actually a thing for initialization purposes, but would have to carefully read the code to be sure
   - Not sure I want to put it on the same condition/error message, but was just making a minimal change to make mypy happier and ensure that there were no other cases which could return None
- Changing one list to a tuple, because the type hint now allows the checker to see an inconsistency (not a runtime problem, to be clear)
   - This is one of the things I discussed above, I've type hinted some methods more narrowly than the implementation actually allows for some places... may relax that, but losing "you need precisely 4 floats" is not ideal...

I'd want to think a little more about this change, in particular whether the return type should be `Figure` or `FigureBase` (and if the former, how we _know_ such/how we instruct the type checker of such, which could be by noop-at-runtime casts or by `#type: ignore` comments, or by `isinstance` checks)

The `<Type>Base` pattern we use in several places for key object hierarchies is a bit hairy for type checking.

For internal code, where we may actually wish to keep `FigureBase` separate from `Figure` and not guarantee that all methods from `Figure` are present (and ensure they aren't used) it is actually quite good. But for user code where 99% of users will get `Figure` but we can't quite guarantee it, it is tricky to spot that balance.

bschnurr commented:
I see. yes better to leave this out. great work by the way

tacaswell commented on lib/matplotlib/pyplot.py:
```diff
@@ -895,24 +965,24 @@ def gcf():
 
 
-def get_current_fig_manager():
+def get_current_fig_manager() -> FigureManagerBase | None:
```
In the future we should probably tighten this up eventually, as while the base `Canvas` may have `obj.manager` as `None`, but if it is coming out of `gcf()` then it is a figure that pyplot is managing and thus we know that manager must not be `None`.

Is this easy to express?  I think this may be a good use of `cast(FigureManagerBase, ret)`?

tacaswell commented on tools/check_typehints.py:
```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env python
```
What is this script used for?

ksunden replied:
Added a docstring. 

Short version is custom checking that since I figured out configuring `stubtest` are _mostly_ not needed, but will keep around for at least the initial roll out stages as an additional (manual, not CI) test to run as a sanity check/for things stubtest currently ignores.

tacaswell reviewed (approved):
I reviewed all of the changes to .py files but trusted that the updates to the pyi files are correct because the tooling is passing.

The funnyness around both `plt.figure` and `plt.get_current_figure_manager` are probably things we should tighten up, but happy to defer that to a follow on PR.

A documentation comment about `check_typehints` would also be good but won't block over that.

ksunden commented:
Test failure is due to new method added on `main` (ecdf, #24728), I'll rebase again and add that in. (Once merged, this will be caught on the individual PR)

ksunden commented:
Post rebase test failure on Azure is simply the stochastic `test_blit` that is a known mildly flaky test

jklymak reviewed (approved):
Obviously I didn't check all the stubs, but the py files seems reasonable to me (only bug change is `__init__.py` now has an `__all__`.  The rest of this is internal and can be refined...

QuLogic commented on requirements/testing/mypy.txt:
```diff
@@ -0,0 +1,27 @@
+# Extra pip requirements for the GitHub Actions flake8 build
```
```suggestion
# Extra pip requirements for the GitHub Actions mypy build
```

QuLogic commented on tools/check_typehints.py:
```diff
@@ -0,0 +1,319 @@
+  - Inheritance makes the set arithmetic harder when only loading AST
+  - Attributes also make it more complicated when defined in init
+- Functions type hinted with ``overload`` are ignored for argument checking
+  - Usually this means the implementation is is less strict in signature but will raise
```
```suggestion
  - Usually this means the implementation is less strict in signature but will raise
```

QuLogic commented on lib/matplotlib/pyplot.py:
```diff
@@ -2050,16 +2177,19 @@ def thetagrids(angles=None, labels=None, fmt=None, **kwargs):
 
 @_api.deprecated("3.7", pending=True)
-def get_plot_commands():
+def get_plot_commands() -> Generator[str, None, None]:
```
#25580

ksunden commented on lib/matplotlib/pyplot.py:
```diff
@@ -1115,7 +1190,7 @@ def axes(arg=None, **kwargs):
 
 
-def delaxes(ax=None):
+def delaxes(ax: matplotlib.axes.Axes | None = None) -> None:
```
See https://github.com/matplotlib/matplotlib/pull/24976#discussion_r1093784988

Short answer is that its unclear to me, too. But sphinx has trouble sometimes because of multiple classes just called `Axes`. I think the differentiating factor that actually matters is that it can properly resolve a type hint that is `Axes`, but cannot resolve it out of a union/composite type (e.g. `Axes | None` or `tuple[Figure, Axes]`)

Only one instance of using `matplotlib.axes.Axes` fully spelled out exists that is alone, and that could very well have just been me spotting that one when changing the ones that didn't work.

(`axes.Axes` also doesn't work because `axes` is a defined function in this file, and doing `maxes.Axes` with `import matplotlib.axes as maxes` I think would also generate sphinx problems, but perhaps the variety that could at least be silenced)

QuLogic replied:
Aha, I thought I had asked this before, but couldn't find it in all the comments that GitHub minimized.

ksunden commented on .github/workflows/mypy-stubtest.yml:
```diff
@@ -0,0 +1,31 @@
+
+      - name: Run mypy stubtest
+        run: |
+          set -o pipefail
```
Ah, yeah, That was just a copy/paste deal (/mildly hopeful I could have done reviewdog to get inline comments on this format as well, but don't think that info is present to even write my own parser)

gramster commented:
> @oscargus I was aware, they were a valuable resource in this exercise, though are not fully complete nor correct (I believe many of the hints in that repo are extracted from inspecting running code, but there are many cases where the type is wider than it saw, e.g. zorder can be any float, but is most often set to int)

I created those stubs; they were a mix of using the docstrings and monkeytype execution traces.

tacaswell commented:
🎉 Thank you @ksunden for getting this done!


I expect there is going to be a bunch more iteration on these as they actually get used (initially by people running of the main branch and then a much wider audience when we get 3.8 out).

gramster commented:
> As I've just looked into typing for another project: how does this work out with pyright? My experience from the other project is that pyright is somewhat easier to make happy... Not that it should be an argument to select one, but it would be nice if it worked well with both (as pyright is more often used for type hinting in IDEs if I am not wrong).

You are not wrong; pyright is used in many editors that use language server protocol (and provides the main type evaluation engine for pylance in Visual Studio Code). Use in editors requires speed and laziness, which pyright has. Pyright can do some amount of type inference, but type inference in Python is hard, which is why we (Microsoft) love it when packages have type annotations or stubs :-)

hauntsaninja commented:
@ksunden author of stubtest here, please open an issue on the mypy repo and tag me if there's anything I can help with!
